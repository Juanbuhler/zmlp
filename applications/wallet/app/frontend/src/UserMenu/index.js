import { useRef, useState } from 'react'
import PropTypes from 'prop-types'

import userShape from '../User/shape'

import { colors, typography, spacing, constants, zIndex } from '../Styles'

import { onBlur as onBlurHelper } from './helpers'

const SIZE = 28

const UserMenu = ({
  user: { firstName, lastName, email },
  logout,
}) => {
  const container = useRef(null)

  const [isMenuOpen, setMenuOpen] = useState(false)

  const onBlur = onBlurHelper({ container, setMenuOpen })

  return (
    <div ref={container} css={{ position: 'relative' }}>
      <button
        type="button"
        onBlur={onBlur}
        onClick={() => setMenuOpen(!isMenuOpen)}
        css={{
          border: 0,
          margin: 0,
          padding: 0,
          width: SIZE,
          height: SIZE,
          borderRadius: SIZE,
          color: colors.rocks.white,
          backgroundColor: colors.rocks.charcoal,
          fontWeight: typography.weight.bold,
          ':hover': {
            cursor: 'pointer',
          },
        }}>
        {`${firstName[0]}${lastName[0]}`}
      </button>
      {isMenuOpen && (
        <div
          css={{
            position: 'absolute',
            zIndex: zIndex.reset,
            backgroundColor: colors.rocks.iron,
            top: SIZE + spacing.base,
            right: 0,
            borderRadius: constants.borderRadius.small,
            boxShadow: constants.boxShadows.menu,
          }}>
          <div
            css={{
              padding: spacing.normal,
              borderBottom: constants.borders.separator,
            }}>
            <div css={{ fontWeight: typography.weight.bold }}>
              {`${firstName} ${lastName}`}
            </div>
            <div>{email}</div>
          </div>
          <div
            css={{
              padding: spacing.normal,
              display: 'flex',
              flexDirection: 'column',
            }}>
            <button
              type="button"
              onBlur={onBlur}
              onClick={logout}
              css={{
                border: 0,
                margin: 0,
                padding: 0,
                backgroundColor: 'transparent',
                fontSize: 'inherit',
                textAlign: 'inherit',
                color: 'inherit',
                ':hover': {
                  color: colors.green2,
                  cursor: 'pointer',
                },
              }}>
              Sign Out
            </button>
          </div>
        </div>
      )}
    </div>
  )
}

UserMenu.propTypes = {
  user: PropTypes.shape(userShape).isRequired,
  logout: PropTypes.func.isRequired,
}

export default UserMenu
