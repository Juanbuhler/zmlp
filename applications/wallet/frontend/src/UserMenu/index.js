import PropTypes from 'prop-types'
import Link from 'next/link'

import userShape from '../User/shape'

import { colors, typography, spacing, constants } from '../Styles'

import ChevronSvg from '../Icons/chevron.svg'

import Menu from '../Menu'
import Button, { VARIANTS } from '../Button'

const SIZE = 28

const UserMenu = ({ user: { firstName, lastName, email }, logout }) => {
  return (
    <div css={{ marginRight: -spacing.moderate }}>
      <Menu
        open="left"
        button={({ onBlur, onClick, isMenuOpen }) => (
          <Button
            aria-label="Open user menu"
            variant={VARIANTS.MENU}
            style={{
              ...(isMenuOpen
                ? { backgroundColor: colors.structure.smoke }
                : {}),
              ':hover': {
                cursor: 'pointer',
              },
            }}
            onBlur={onBlur}
            onClick={onClick}
          >
            <div css={{ display: 'flex', alignItems: 'center' }}>
              <div
                css={{
                  display: 'flex',
                  justifyContent: 'center',
                  alignItems: 'center',
                  border: 0,
                  padding: 0,
                  width: SIZE,
                  height: SIZE,
                  borderRadius: SIZE,
                  color: isMenuOpen
                    ? colors.structure.white
                    : colors.structure.lead,
                  backgroundColor: isMenuOpen
                    ? colors.structure.lead
                    : colors.structure.steel,
                  fontWeight: typography.weight.bold,
                }}
              >
                {`${firstName ? firstName[0] : ''}${
                  lastName ? lastName[0] : ''
                }`}
              </div>
              <ChevronSvg
                height={constants.icons.regular}
                color={colors.structure.steel}
                css={{
                  marginLeft: spacing.base,
                  transform: `${isMenuOpen ? 'rotate(-180deg)' : ''}`,
                }}
              />
            </div>
          </Button>
        )}
      >
        {({ onBlur, onClick }) => (
          <div>
            <div
              css={{
                padding: spacing.normal,
                borderBottom: constants.borders.regular.zinc,
              }}
            >
              <div css={{ fontWeight: typography.weight.bold }}>
                {`${firstName} ${lastName}`}
              </div>
              <div>{email}</div>
            </div>
            <ul css={{ borderBottom: constants.borders.regular.zinc }}>
              <li>
                <Link href="/account" passHref>
                  <Button
                    variant={VARIANTS.MENU_ITEM}
                    onBlur={onBlur}
                    onClick={onClick}
                  >
                    Manage Account
                  </Button>
                </Link>
              </li>
              <li>
                <Button
                  href="mailto:zvi-console-support@zorroa.com"
                  target="_blank"
                  variant={VARIANTS.MENU_ITEM}
                  onBlur={onBlur}
                  onClick={(event) => {
                    onClick(event)
                  }}
                >
                  Contact Support
                </Button>
              </li>
              <li>
                <Button
                  href="https://shipright.community/zorroa"
                  target="_blank"
                  variant={VARIANTS.MENU_ITEM}
                  onBlur={onBlur}
                  onClick={onClick}
                >
                  Submit Feedback
                </Button>
              </li>
            </ul>
            <ul>
              <li>
                <Button
                  variant={VARIANTS.MENU_ITEM}
                  onBlur={onBlur}
                  onClick={() => logout({ redirectUrl: '/', redirectAs: '/' })}
                >
                  Sign Out
                </Button>
              </li>
            </ul>
          </div>
        )}
      </Menu>
    </div>
  )
}

UserMenu.propTypes = {
  user: PropTypes.shape(userShape).isRequired,
  logout: PropTypes.func.isRequired,
}

export default UserMenu
