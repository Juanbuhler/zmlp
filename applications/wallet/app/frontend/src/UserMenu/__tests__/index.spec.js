import TestRenderer, { act } from 'react-test-renderer'

import mockUser from '../../User/__mocks__/user'

import UserMenu from '..'

const noop = () => () => {}

describe('<UserMenu />', () => {
  it('should render properly', () => {
    const component = TestRenderer.create(
      <UserMenu user={mockUser} logout={noop} />,
    )

    expect(component.toJSON()).toMatchSnapshot()

    act(() => {
      component.root
        .findByProps({ children: 'Jane Doe' })
        .props.onClick({ preventDefault: noop })
    })

    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should render properly without firstName/lastName', () => {
    const component = TestRenderer.create(
      <UserMenu
        user={{ ...mockUser, firstName: '', lastName: '' }}
        logout={noop}
      />,
    )

    expect(component.toJSON()).toMatchSnapshot()
  })
})
