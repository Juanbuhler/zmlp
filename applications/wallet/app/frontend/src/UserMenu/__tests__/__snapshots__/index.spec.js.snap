// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserMenu /> should render properly 1`] = `
.emotion-5 {
  margin-right: -12px;
}

.emotion-4 {
  position: relative;
}

.emotion-3 {
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-radius: 2px;
  margin: 0;
  padding: 12px 32px;
  white-space: nowrap;
  font-size: 14px;
  line-height: 16px;
  font-weight: 500;
  border: none;
  cursor: pointer;
  background-color: transparent;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  padding: 4px 8px;
  color: #82be3c;
  font-weight: 400;
  height: 100%;
  false: #3b3b3b;
}

.emotion-3:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-3:hover {
  background-color: #3b3b3b;
}

.emotion-3:hover {
  cursor: pointer;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 0;
  margin: 0;
  padding: 0;
  width: 28px;
  height: 28px;
  border-radius: 28px;
  color: #252525;
  background-color: #808080;
  font-weight: 700;
}

.emotion-1 {
  margin-left: 8px;
}

<div
  className="emotion-5"
>
  <div
    className="emotion-4"
  >
    <button
      aria-label="Open user menu"
      className="emotion-3"
      onBlur={[Function]}
      onClick={[Function]}
      type="button"
    >
      <div
        className="emotion-2"
      >
        <div
          className="emotion-0"
        >
          JD
        </div>
        <svg
          className="emotion-1"
          color="#808080"
          data-file-name="chevron.svg"
          width={20}
        />
      </div>
    </button>
  </div>
</div>
`;

exports[`<UserMenu /> should render properly 2`] = `
.emotion-9 {
  margin-right: -12px;
}

.emotion-8 {
  position: relative;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-3 {
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-radius: 2px;
  margin: 0;
  padding: 12px 32px;
  white-space: nowrap;
  font-size: 14px;
  line-height: 16px;
  font-weight: 500;
  border: none;
  cursor: pointer;
  background-color: transparent;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  padding: 4px 8px;
  color: #82be3c;
  font-weight: 400;
  height: 100%;
  background-color: #3b3b3b;
}

.emotion-3:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-3:hover {
  background-color: #3b3b3b;
}

.emotion-3:hover {
  cursor: pointer;
}

.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 0;
  margin: 0;
  padding: 0;
  width: 28px;
  height: 28px;
  border-radius: 28px;
  color: #ffffff;
  background-color: #252525;
  font-weight: 700;
}

.emotion-1 {
  margin-left: 8px;
  -webkit-transform: rotate(-180deg);
  -ms-transform: rotate(-180deg);
  transform: rotate(-180deg);
}

.emotion-7 {
  position: absolute;
  z-index: 1;
  top: 100%;
  right: 0;
  background-color: #808080;
  border-radius: 2px;
  box-shadow: 0 4px 7px 0 #000000;
  width: 200px;
}

.emotion-7 ul {
  list-style-type: none;
  padding: 0;
  padding-top: 4px;
  padding-bottom: 8px;
  margin: 0;
}

.emotion-7 li {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-5 {
  padding: 16px;
  padding-bottom: 24px;
  border-bottom: 1px solid #9b9b9b;
}

.emotion-4 {
  font-weight: 700;
}

.emotion-6 {
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-radius: 2px;
  margin: 0;
  padding: 12px 32px;
  white-space: nowrap;
  font-size: 14px;
  line-height: 16px;
  font-weight: 500;
  border: none;
  cursor: pointer;
  background-color: transparent;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  color: #ffffff;
  padding: 8px 16px;
  font-weight: 400;
  border-radius: 0;
  padding: 8 0;
  border-bottom: 1px solid transparent;
}

.emotion-6:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-6:hover {
  background-color: #4a4a4a;
}

<div
  className="emotion-9"
>
  <div
    className="emotion-8"
  >
    <button
      aria-label="Open user menu"
      className="emotion-3"
      onBlur={[Function]}
      onClick={[Function]}
      type="button"
    >
      <div
        className="emotion-2"
      >
        <div
          className="emotion-0"
        >
          JD
        </div>
        <svg
          className="emotion-1"
          color="#808080"
          data-file-name="chevron.svg"
          width={20}
        />
      </div>
    </button>
    <div
      className="emotion-7"
    >
      <div>
        <div
          className="emotion-5"
        >
          <div
            className="emotion-4"
          >
            Jane Doe
          </div>
          <div>
            jane.doe@zorroa.com
          </div>
        </div>
        <ul>
          <li>
            <button
              className="emotion-6"
              onBlur={[Function]}
              onClick={[Function]}
              type="button"
            >
              Sign Out
            </button>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;

exports[`<UserMenu /> should render properly without firstName/lastName 1`] = `
.emotion-5 {
  margin-right: -12px;
}

.emotion-4 {
  position: relative;
}

.emotion-3 {
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-radius: 2px;
  margin: 0;
  padding: 12px 32px;
  white-space: nowrap;
  font-size: 14px;
  line-height: 16px;
  font-weight: 500;
  border: none;
  cursor: pointer;
  background-color: transparent;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  padding: 4px 8px;
  color: #82be3c;
  font-weight: 400;
  height: 100%;
  false: #3b3b3b;
}

.emotion-3:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-3:hover {
  background-color: #3b3b3b;
}

.emotion-3:hover {
  cursor: pointer;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 0;
  margin: 0;
  padding: 0;
  width: 28px;
  height: 28px;
  border-radius: 28px;
  color: #252525;
  background-color: #808080;
  font-weight: 700;
}

.emotion-1 {
  margin-left: 8px;
}

<div
  className="emotion-5"
>
  <div
    className="emotion-4"
  >
    <button
      aria-label="Open user menu"
      className="emotion-3"
      onBlur={[Function]}
      onClick={[Function]}
      type="button"
    >
      <div
        className="emotion-2"
      >
        <div
          className="emotion-0"
        >
          OP
        </div>
        <svg
          className="emotion-1"
          color="#808080"
          data-file-name="chevron.svg"
          width={20}
        />
      </div>
    </button>
  </div>
</div>
`;
